segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -187,26 +257,64 @@ public void initialize() throws Exception {
             val finalTruncateOffsets = new HashMap<String, Long>();
             val finalFirstChunkStartsAtOffsets = new HashMap<String, Long>();
 
+            // Keep track of already processed records.
+            val visitedRecords = new HashSet<SystemJournalRecord>();
+
             // Step 1: Create metadata records for system segments from latest snapshot.
-            val epochToStart = applyLatestSnapshot(txn, chunkStartOffsets);
+            val latestSnapshot = findLatestSnapshot();
+
+            applySystemSnapshotRecord(txn, chunkStartOffsets, latestSnapshot);
 
             // Step 2: For each epoch, find the corresponding system journal files, process them and apply operations recorded.
-            applySystemLogOperations(txn, epochToStart, chunkStartOffsets, finalTruncateOffsets, finalFirstChunkStartsAtOffsets);
+            applySystemLogOperations(txn, latestSnapshot, chunkStartOffsets, finalTruncateOffsets, finalFirstChunkStartsAtOffsets, visitedRecords);
 
             // Step 3: Adjust the length of the last chunk.
-            adjustLastChunkLengths(txn);
+            if (config.isLazyCommitEnabled()) {
+                adjustLastChunkLengths(txn);
+            }
 
             // Step 4: Apply the truncate offsets.
             applyFinalTruncateOffsets(txn, finalTruncateOffsets, finalFirstChunkStartsAtOffsets);
 
             // Step 5: Validate and save a snapshot.
-            validateAndSaveSnapshot(txn);
+            boolean saved = validateAndSaveSnapshot(txn, true, config.isSelfCheckEnabled());
+            if (saved) {
+                writeSnapshotInfo(lastSavedSystemSnapshotId);
+            }
 
-            // Step 5: Finally commit all data.
+            // Step 6: Check invariants.
+            Preconditions.checkState(currentFileIndex == 0, "currentFileIndex must be zero");
+            Preconditions.checkState(systemJournalOffset == 0, "systemJournalOffset must be zero");
+            Preconditions.checkState(newChunkRequired, "newChunkRequired must be true");
+            if (saved) {
+                Preconditions.checkState(lastSavedSystemSnapshot != null, "lastSavedSystemSnapshot must be initialized");
+            }
+
+            // Step 7: Finally commit all data.
             return txn.commit(true, true);
         }
     }
 
+    @SneakyThrows
```

Comment:
What does it throw? You catch exceptions below.

Reply:
Fixed.